### ISSUE TITLE
Incorrect resampling with multiple cores

### ISSUE THREAD

**Issue by nonplus** [viliusle/Hermite-resize#4]
When I resample an 112x113 image to 40x40, I end up with half the image missing on my Mac (8 cores).  Resampling the image using `resample_simple` produces the correct results.

**Source image (112x113):**

![rainbow 112x113](https://cloud.githubusercontent.com/assets/1023803/21156469/8d4b5e5c-c13b-11e6-82d5-8a8297222066.png)

**Incorrect result using `resample(canvas, 40, 40, true)`:**

![rainbow 40x40 8 cores](https://cloud.githubusercontent.com/assets/1023803/21156498/a6e06cae-c13b-11e6-81a0-dacb9e57610d.png)

**Correct result using `resample_single(canvas, 40, 40, true)`:**

![rainbow 40x40 single](https://cloud.githubusercontent.com/assets/1023803/21156513/b47c2cc2-c13b-11e6-82df-600eb8ed11bb.png)

FWIW, doing a `resample(canvas, 50, 50, true)` on the same source images produces a correct result; as does resampling larger images to 40x40.  Other resample values for this source images produce wrong results (e.g. `39x39`, `32x32`, etc.)  Probably a rounding error somewhere in there...

**Comment by viliusle** [comment]
many cores, small image... i will check it soon.

**Comment by prototype23** [comment]
I had reported the same issue from a mac user, where some times the image where resized with these black gaps.

**Comment by prototype23** [comment]
The same happens in Microsoft's Edge Browser at Windows with bigger images.

![capture](https://cloud.githubusercontent.com/assets/14275025/21319732/17442cf8-c617-11e6-97a7-11cb3b900fbc.JPG)
![capture2](https://cloud.githubusercontent.com/assets/14275025/21319731/17431462-c617-11e6-8c6a-8da1d26b1af3.JPG)
![capture3](https://cloud.githubusercontent.com/assets/14275025/21319733/17459dcc-c617-11e6-8784-3639db223dd7.JPG)

Initial image canvas size was around 700px and target to 512px. Screenshots above are thumbnails.

**Comment by viliusle** [comment]
I see problem. Checking what is wrong.

**Comment by viliusle** [comment]
all related issues fixed in v2.2

**Comment by prototype23** [comment]
Thanks for the fix and the new features.

In the meanwhile, to overcome the problem i had created a custom worker using the single version. Since you have fixed the problem i have to ask you one more question. 

What are the advantages of using the multi-core version?

Merry Christmas!!

**Comment by viliusle** [comment]
In general multi-core version:

1. (+) work can be done in background, browser is not locked during processing, user can do other tasks, if image is very very big.
2. (+) process can take long time, user will not get alert asking to stop/continue script.
3. (+) also work can be done faster by using all cpu cores. 10 cores = 10 times faster process in theory. 
4. (-) In practice, you must add extra time for creating worker, splitting and combining data.

Using only 1 worker can be done (like you did), but you gain only #1, #2 but not #3 advantages.  And you wasting time on #4.

In this project resizing small image using workers will be slower (1 core f-tion will finish very fast). resizing big images will be similar in speed (because of #4) and only resizing very big images will be faster using multi-core function.

In short, multi-core version was added because of: there was interest from other people in multi-core, challenge, #1. But #4 is really big problem and usually #4 loss = #3 gain (for image processing, unless you will do really long calculations with small data)

In general simple (no multi-core) function works really fast and it can be used without any big disadvantages.

**Comment by prototype23** [comment]
Thanks for the very detailed answer.  

BTW i have no experience with thread sharing. So in theory, when we create a web worker, browser creates a new thread on the less active cpu and process the calculations there?

**Comment by viliusle** [comment]
yes. Most important, if you have cpu with 16 cores, you can create 16 web workers and they will do its job in same time. Single core version would take 16x more time, since every taks will be executed only 1 in time. In theory of course.
